Userland/Utilities/test-js.cpp

```diff
@@ -713,6 +719,7 @@ int main(int argc, char** argv)
     args_parser.add_option(print_times, "Show duration of each test", "show-time", 't');
     args_parser.add_option(collect_on_every_allocation, "Collect garbage after every allocation", "collect-often", 'g');
     args_parser.add_option(test262_parser_tests, "Run test262 parser tests", "test262-parser-tests", 0);
+    args_parser.add_option(format_errors_for_ci, "Output error messages that CI understands", "ci-output", 'c');
```

Comment:
This is GitHub Actions specific, so please reflect that in the description, flag name and/or variable name :)

Reply:
Mmk, if I keep this change for test-js I'll make sure that's noted.

I'm considering whether to just make the shell script do the fancy notifications, or if it's worth the "ugliness" to embed some knowledge of github actions into the test apps?

```sh
#!/bin/sh

echo "==== Running Tests on SerenityOS ===="
test-js
if [ $? -ne 0];
    echo "::error file=test-js:: Test returned non-zero exit code, refer to logs"
fi

echo "==== Done running tests, shutting down.... ===="
shutdown -n
```

Reply:
(Yes, I know this was 4 days ago, consider this a nudge)
Friendly reminder that the equivalent in /bin/Shell would be this:
```sh
echo "..."
if test-js {}
else {
    echo "::error file=blah:: 'if' needs negation"
}

echo "yoinks"
shutdown -n
```

Reply:
TIL:
```sh
$ not ls 2>&1 >/dev/null; echo $?
1

$ pacman -Qo /usr/bin/not
/usr/bin/not is owned by llvm 11.0.1-1
```

Reply:
I figured you'd chime in with the correct form eventually :p. I think that doing the shell thing is the way to go. But ideally, I'd want to create a list of "all tests" and just do this for all tests. Opened up 2 shell issues that either point out deficiencies in the shell's for loops/if statements, or in my reading of the docs before coffee as a result of messing around with this.

Reply:
Something like the shell equivalent of this python snippet:

```python
print("Starting the tests yo")
all_tests=[ ??? ] # tbd method for aggregating tests

for idx, test in enumerate(all_tests):
    print("Running test {}/{} -- {}", idx, len(all_tests), test)
    if magic_exec(test):
        print("::info file={0}:: {0} passed!", test)
    else:
        print("::error file={0}:: {0} failed!", test)
        
print("done with the tests")
magic_exec("shutdown -n")
```

Reply:
hmmm...indices in a for loop :thinking:
I guess you'll have to go the hard way for that, `i=0` before the loop and `i=$(expr $i + 1)` inside the loop.
alternatively implement `-n` for `cat` then use `match` to split the index (I've had thoughts about allowing `for match pattern as (names) in expr` and something similar for normal variable declarations, but I don't have anything concrete).

the rest should be fine though:
```sh
echo Starting the tests yoooooo
all_tests=( ??? )
# (list count doesn't exist yet, but you can cheese it:)
count=$(printf %s\n $all_tests | wc -l)
idx=0
for test in $all_tests {
    echo "Running test $idx/$count -- $test"
    if $test {
		echo whatever $test $test $test
	}
    else {
        echo on noes we broke it
    }
    idx=$(expr $idx + 1)
}
echo Done with the toasts
shutdown -n
```

Reply:
Either I'm missing something (seems to be a trend ;) ) or there's some more shell bugs going on here... This crashes the shell with a segfault, even with only one test. Also the spacing on the count is super weird.

Note this is text mode though.

```sh
#!/bin/sh

echo "==== Running Tests on SerenityOS ===="

# TODO: Enable all tests when shell stops crashing :(
#IFS='\n'
#usr_tests_tests=($(find /usr/Tests -type f))
#unset IFS
#all_tests=("test-js" "test-web" "test-pthread" "test-crypto" "test-compress" $usr_tests_tests)
all_tests=("test-js")

# TODO: List size checking when?
count=$(printf %s\n $all_tests | wc -l)
idx=1

for test_name in $all_tests {
    echo "Running test $idx/$count -- $test_name"
    if $test_name {
        echo "::info file=$test_name:: $test_name passed!"
    } else {
        echo "::error file=$test_name:: $test_name returned non-zero exit code, check logs!"
    }
    idx=$(expr $idx + 1)
}

echo "==== Done running tests, shutting down.... ===="
shutdown -n
```

![spacing](https://user-images.githubusercontent.com/8388494/104858770-dd142100-58de-11eb-8577-6d5c9468771a.png)
![crashy-crash](https://user-images.githubusercontent.com/8388494/104858772-ddacb780-58de-11eb-85f1-e092b46ade62.png)

Reply:
Wait, wat is all this?
```
[expr(55:55)]: expr(55:55) Should already die
Shell(56): Job entry "expr 1 + 1" deleted in 251 ms
[Shell(56:56)]: File is not an ELF file.
[Shell(56:56)]: exec(/bin/): File has invalid ELF header
Shell(56): execvpe() failed on attempt (/bin/) with Exec format error
Shell(56): execvp() about to return -1 with errno=8
[Shell(56:56)]: Shell(56:56) Should already die
```

Reply:
That's some weird looking error

Reply:
Ohhh
Our `wc` is emitting extra spaces for some reason

Reply:
I'm not seeing a segfault, but things are really weird alright

Reply:
Something is fishy here, why is a subshell also being notified of these job completions, and why does it even have access to them? so many bugs

Also, I don't think we have support for nested lists yet, and you can't run a string like in bash (it won't split, this was a design decision - and I still think it's a good one).
you _can_ get around that by doing something really weird:
```sh
IFS=" "
if $($(echo $test_string)) { } else { }
IFS="\n"
```

Reply:
Let's take a step back and fix bugs instead of making workarounds, silly me.

Reply:
Okay, so with #4996, this works:
```sh
#!/bin/sh

echo "==== Running Tests on SerenityOS ===="

run() {
    test_cmd=($*)
    echo "Running test $idx/$count -- $test_cmd"
    if $test_cmd {
        echo "::info file=$test_cmd:: $test_cmd passed!"
    } else {
        echo "::error file=$test_cmd:: $test_cmd returned non-zero exit code, check logs!"
    }
    idx=$(expr $idx + 1)
}

# TODO: It'd be nice to have a list+list op (as opposed to nest-on-in-another)
system_tests=(test-js (test-crypto bigint -t)) #("test-js" "test-web" "test-pthread" "test-crypto" "test-compress" $usr_tests_tests)
usr_tests_tests=$(find /usr/Tests -type f)
all_tests=($system_tests $usr_tests_tests)

count=0
for list in $all_tests {
    for $list { count=$(expr $count + 1) }
}

idx=1
for list in $all_tests {
    for $list { run $it }
}

echo "==== Done running tests, shutting down.... ===="
shutdown -n
```

On one hand, I like the fact that lists aren't magically flattened until you pass them to a command, and on the other hand, we get these nested for loops...:thinking:

Reply:
Looks like that does work, yeah. The loop to count all the files runs... very very slowly on my setup (No KVM or windows hyperv accel). Lotta shells to spawn to count all 80 tests. The tests don't all work though. Missing some way to either

```
$(find /usr/Tests -type f | grep -v Kernel)
```
or
```
$(find /usr/Tests -not -path /usr/Tests/Kernel -type f)
```
Now :D Not sure which would be easier to implement... probably `grep -v`

Reply:
Yeah, I imagine it'd be quite slow, really should figure out the list length thing, but I've absolutely hated all the attempts at "variable substitution" so far

If you don't feel like grep -v, you could try using a shell function (and probably finding a few more bugs)

Reply:
you're not wrong about the bugs :D

With my commits from #5089 applied... 

Runs ~80 tests, but some of them are not tests but instead proof of concept ad hoc feature integration programs that no longer work:
```sh
system_tests=(test-js (test-crypto bigint -t)) #("test-js" "test-web" "test-pthread" "test-crypto" "test-compress" $usr_tests_tests)
usr_tests_tests=$(find /usr/Tests -type f)
all_tests=($system_tests $usr_tests_tests)
```
Segmentation fault in the Shell:
```sh
system_tests=(test-js (test-crypto bigint -t)) #("test-js" "test-web" "test-pthread" "test-crypto" "test-compress" $usr_tests_tests)
usr_tests_tests=$(find /usr/Tests -type f | grep -v Kernel)
all_tests=($system_tests $usr_tests_tests)
```

Runs two tests, but the usr_test_tests variable refuses to turn into an array no matter what I try (Shell inexperience showing...):
```sh
system_tests=(test-js (test-crypto bigint -t)) #("test-js" "test-web" "test-pthread" "test-crypto" "test-compress" $usr_tests_tests)
usr_tests_tests=($(find /usr/Tests -type f | grep -v Kernel))
all_tests=($system_tests $usr_tests_tests)
```

Reply:
Er ok I'm confused. Adding `#define ENSURE_WAITID_ONCE` at the top of Shell.cpp fixes the crash on the second one. but.. there's already an `#ifdef __serenity__ #define ENSURE_WAITID_ONCE` ...?

Reply:
(while I rebuild things)

a segfault _inside_ waitid? that doesn't sound like a shell bug (wait.cpp:100 is the syscall)

> usr_test_tests variable refuses to turn into an array

it already is a list though? `$()` will always be a list, no need to put it in another list

Reply:
Looks like the shell/kernel weirdness is stable enough to run the tests consistently on my machine